### PR TITLE
Major modernization.

### PR DESCRIPTION
**Pull request by mralexgray**
Please see changes and commit messages and let me know if you have any questions/trepidation.

screen shot 2015-04-25 at 3 18 54 pm

### CODE REVIEW

**Comment by markcarver** [comment]
Wow @mralexgray :) This is awesome! Certainly quite extensive. I'm not entirely sure when I can get to this to review though. If not this week, ping here or in freenode IRC (#cocoadialog) to remind me.

**Comment by mralexgray** [comment]
@markcarver  sorry this was so convoluted.  i would really like to use this, as a framework, in other cocoa projects... which was the point of the original refactor.  this was a while back, now.  if you'd  like, add any comments to the individual commits, and I can either rework them, eliminate them, or we can try to merge them.

**Comment by markcarver** [comment]
@mralexgray, these are indeed a lot of changes. I really haven't worked on this project for a couple years (even though, yes I did the refactor last time).

Personally, I would say let's go for it since the current code is based on an older xcode version.

@mstratman, thoughts?

**Comment by mstratman** [comment]
I don't have time to do a thoughtful review, and without a test suite it'll require a bit more manual testing. But if it seems to still be working okay, go for it.

File: cocoaDialog.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -529,84 +597,137 @@
...
+			files = (
+			);
+			inputPaths = (
+				"$(BUILT_PRODUCTS_DIR)/$(PROJECT_NAME).app/Contents/MacOS/$(PROJECT_NAME)",
+			);
+			outputPaths = (
+				/usr/local/bin,
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/zsh;
```
Comment (markcarver) [inline cocoaDialog.xcodeproj/project.pbxproj]: No everyone will have `zsh` (I don't). It's better if we stick with `sh`.

**Comment by markcarver** [comment]
@mralexgray, just went through all the commits rather quickly. Everything seems rather straight forward (AFAICT, considering it touches nearly every single file). I've made the one comment about `zsh` and would prefer that be changed.

---

@mstratman, same situation here. Totally agree about needing test suites for this. I had a shell script, at one point, that went through and semi-automatically tested everything. I'll have to dig it out of my old laptop's archive dmg.

FWIW, this project has been really stale (3.x has been in "beta" for years). I think any change to the code on GH (good or bad) would be rather welcome at this point considering I was the one that originated the rewrite of 3.x.

I would be ok merging under the assumption that @mralexgray is willing to provide any subsequent PRs to fix stuff and/or add tests (e.g. travis ci?).

**Comment by markcarver** [comment]
For reference: http://docs.travis-ci.com/user/languages/objective-c/

**Comment by markcarver** [comment]
Ah, no need to dig through my old laptop for that testing shell script, I created a gist: https://gist.github.com/markcarver/7a593f2a000c33e07142

**Comment by mralexgray** [comment]
i'm continuing work as we speak.  I am about to use the standard shell version in a script, so that will help confirm the classic usage is still functional.  will make changes to the zsh thing and report back shortly.

**Comment by mralexgray** [comment]
@mstratman  can someone tell me what notify is supposed to do? even a fresh checkout of master branch does nothing when i try to run it with notify.  can i just axe notify + growl?

**Comment by mralexgray** [comment]
also, play around with `Examples/examples.sh`  it is the beginning of a testable verification harness.

**Comment by mralexgray** [comment]
@markcarver thanks for that script... i haven't gone through ALL the functions to make sure the test-run is picking up ALL it's rich functionality, but on the surface, everything seems to be working AOK.  secondary verification of this fact would be appreciated!  things seem to be working from the ObjC side too..  so far, I am enable to invoke the various Control classes... have to figure out how I can break the run loop to be able to use more than one "widget" per application run, i anyone has some suggestions...


screen shot 2015-10-03 at 3 23 01 am

**Comment by mralexgray** [comment]
helloooo!?!  i am totally going down a rabbit hole with this.  if enthusiasm wanes, i'll be up sh*t creek.  do you (_someone, anyone_) wanna maybe call this `4.x beta`, and at least but it in an officially tagged branch?  i'm trying hard to keep everything working while i do these sweeping changes.. but i need either feedback, merges, or some positive reinforcement, or I guess I'll just.... well I dunno what i'll do, lol.

**Comment by mstratman** [comment]
If there's a tradeoff to be had, I think functional trumps modern syntax.  If you can deliver both, that's great.  But a few things you said give me pause that perhaps this isn't a stable build. If it is, sure let's add a link to it on the site!

I think it's great you want to update this, and if you do you'll need to take it all the way to 100%. If you're looking to take this to 80% and have somebody else pick up the last 20, that might be  harder to find volunteers.

This is an ENORMOUS pull request, and I will never have the time to review, test, and provide guidance.

Given that nobody is actively developing this, you have a lot of flexibility and freedom to do with it what you'd like, but keep in mind that it'll be on you to ensure any new releases are stable.

I still routinely use the app and bet many others do too, so even though development has slowed, quality can't waver.

**Comment by markcarver** [comment]
> This is an ENORMOUS pull request, and I will never have the time to review, test, and provide guidance.

Agreed. Neither of us have the time to review this, but... it needs to happen and I would prefer to use someone else's time and effort than having to duplicate it myself one day. It would be far easier to for me to get involved if/when this merged (I have a hard time manually reviewing PRs from GH, I know it can be done, but meh).

We never "officially" released the 3.x branch. I would be fine with merging this PR as is. This way @mralexgray can create smaller/newer PRs as needed. Any progress is better than none. We don't have to create a release until we know it's stable. That means adding real/official tests (not scripts).

**Comment by markcarver** [comment]
@mstratman, actually I'm just going to merge this as is. The newer Xcode version is what played a big part for why this PR is so big. I don't think there would have ever been a way around this. We can always revert if needed.

@mralexgray, just an FYI though (moving forward), smaller and more precise PRs if possible. It will be better/easier for all of us involved to review and merge more quickly.

**Comment by mstratman** [comment]
Alright, sounds good. Thanks to both of you for your hard work! :+1:

**Comment by mralexgray** [comment]
Wow, @mstratman , @markcarver , you guys are "my kind" of adventurous!  That's a good thing! To be honest, I rarely expect upstream maintainers to embrace this kind of brutal repackaging - but in this case it is the right thing to do.  I view the functionality of this framework as essential to the platform.  Frankly, Apple should build in it's functionality.  I still see room for a LOT of improvement, so I will just keep chugging, (and testing), but in smaller, more digestible chunks.  Thanks for the active feedback, and lack of fear/resentment to move upwards and onwards!

**Pull request merged**